README
################################################################################

Program		: solr_pager
Version		: 0.2.5
Purpose		: Pager component for SOLR
License		: GNU GPL v3 (see file COPYING)
Author		: Samuel Behan <samuel_._behan_(at)_dob_._sk> (c) 2000-2016
Web		: http://devel.dob.sk/solr_pager
Requirements	: JAVA 1.5, SOLR 1.4+, ANT

################################################################################

---------
- ABOUT -
---------
	solr_pager is a SOLR paging component. It adds paging records to SOLR
  search results, that can be easily transformed using XSLT writer. You will
  not need anymore to bother with slow recursive XSLT paging hacks.
  See PAGING for more info.

-----------------
- BUG REPORTING -
-----------------
	Report bugs or feature request to me (Author).

---------------
- INSTALATION -
---------------
	For use of solr_pager you might need to compile this against yours 
  SOLR version. Altought this tarball contains some recent version of SOLR jars, 
  they might be different and compiled version of solr_pager might not work with 
  yours SOLR. So, you will have to copy your SOLR (located in <webapps>/solr/WEB-INF/lib) 
  jars to libs directory or alter library paths in build.xml.

  After this simply type

   $ ant

  This will create solr_pager.jar in dist directory, that you should copy to yours
  SOLR lib dir (usually <webapps>/solr/WEB-INF/lib).

  Finaly you need to define new component that can SOLR use in solrconfig.xml:

  <searchComponent name="pager" class="sk.dob.search.solr.handler.component.PagerComponent"/>

  And add pager component to request processing:

  <requestHandler name="standard" class="solr.SearchHandler" default="true">

    <!-- ... usual solr requestHandler configurations ... ->

    <!-- use pager as last request processing component -->
    <arr name="last-components">
      <str>pager</str>
    </arr>
  </requestHandler>



----------------------
- SOLR QUERY OPTIONS -
----------------------

	pager		- number of following or preceding page starts to generate
			  default: 0 (disabled)

	pager.pre	- how many of preceding page starts should be generated
			  default: 2

----------
- PAGING -
----------

	Response with start = 75 and rows = 25.

<response>
  <lst name="pager">
    <lst name="pages">		<!-- list of all pages -->
      <int name="2">25</int>
      <int name="3">50</int>
      <int name="4">75</int> <!-- this is actual page -->
      <int name="5">100</int>
      <int name="6">125</int>
      <int name="7">150</int>
      <int name="8">175</int>
      <int name="9">200</int>
      <int name="10">225</int>
      <int name="11">250</int>
    </lst>
    <int name="prev">50</int>	<!-- previous page, with start = 50 -->
    <int name="next">100</int>	<!-- next page, with start = 75 -->
    <int name="last">1225</int>	<!-- last page, with start = 1225 -->
    <int name="actual">4</int> <!-- actual page number -->
    <int name="count">49</int> <!-- count of all pages -->
  </lst>
</response>



ChangeLog
2016-09-01  Sam <samuel_._behan_(at)_dob_._sk>

	* version 0.2.5 for apache-solr-6.1.0

2011-08-25  Sam <samuel_._behan_(at)_dob_._sk>

	* version 0.2.4 for apache-solr-3.3.0

2010-07-22  Sam <samuel_._behan_(at)_dob_._sk>

	* version 0.2.3
	* fixed incorrect last page entry and next/last navigators generation (thanks Marcel S. for reporting problem)

2009-12-06  Sam <samuel_._behan_(at)_dob_._sk>

	* version 0.2.2
	* upgraded libs to stable 1.4 version
	* pager did not worked for standard search handler, so now
		we always request result set from solr
	* added additional checks to prevent working on empty result set

2009-12-06  Sam <samuel_._behan_(at)_dob_._sk>

	* version 0.2.1
	* added first page navigator
	* improved first & last paging, now they appear only if first/last page is not visible

2009-12-04  Sam <samuel_._behan_(at)_dob_._sk>

	* version 0.2.0
	* initial stable version


src/sk/dob/search/solr/handler/component/PagerComponent.java
/** PageComponent.java
 * Program		: solr_pager
 * Version		: 0.2.0
 * Purpose		: Pager component for SOLR
 *			  Provides simple and fast paging for SOLR results
 * License		: GNU GPL v3 (see file COPYING)
 * Author		: Samuel Behan <samkob_(at)_gmail_._com> (c) 2000-2009
 * Web		: http://devel.dob.sk/solr_pager
 * Requirements	: JAVA 1.5, SOLR 1.4, ANT
 */
package sk.dob.search.solr.handler.component;

/* solr imports */
import org.apache.lucene.search.Query;
import org.apache.solr.request.SolrQueryRequest;
import org.apache.solr.handler.component.ResponseBuilder;
import org.apache.solr.handler.component.SearchComponent;
import org.apache.solr.handler.component.ShardRequest;
import org.apache.solr.common.SolrException;
import org.apache.solr.common.params.CommonParams;
import org.apache.solr.common.params.SolrParams;
import org.apache.solr.common.util.NamedList;
import org.apache.solr.common.util.SimpleOrderedMap;


/* lf4j import */
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/* java imports */
import java.io.IOException;
import java.net.URL;


/**
 * PagerComponent
 *   Provides simple and fast paging for SOLR results
 *
 * @version 0.2.0 SamB $
 * @since solr 1.4
 */
public class PagerComponent extends SearchComponent 
{
  public static final String COMPONENT_NAME = "pager";
  public static final String PARAM_PAGER = "pager";
  public static final String PARAM_PAGER_PRE = "pager.pre";
//  private static final Logger log = LoggerFactory.getLogger(PagerComponent.class);

  @Override
  public void prepare(ResponseBuilder rb) throws IOException
  {
	/* notify solr we always need resultset ! */
	if(rb.req.getParams().getInt(PARAM_PAGER, 0) != 0)
		rb.setNeedDocSet(true);
  }

  @Override
  @SuppressWarnings("unchecked")
  public void process(ResponseBuilder rb) throws IOException 
  {
	/* get request params */
	SolrParams par = rb.req.getParams();
	int rows = par.getInt(CommonParams.ROWS, 0);
	int start = par.getInt(CommonParams.START, 0);
	int pages = par.getInt(PARAM_PAGER, 0);
	int pages_pre = par.getInt(PARAM_PAGER_PRE, 2);

	/* neet to work ? */
	if(pages == 0 || rows == 0 || rb == null 
		|| rb.getResults() == null)
		return;

	/* select result list */
	int doc_count = 0; 

	if(rb.getResults().docSet != null)
		doc_count = rb.getResults().docSet.size();
	else
		return;

	/* pager list */
	NamedList lst = new SimpleOrderedMap<Object>();
	NamedList lst2 = new SimpleOrderedMap<Object>();

	/* paging pages */
	int page_count = doc_count / rows;
	int page_actual = start / rows;
	int page_pre = pages_pre;
	int page_post = pages - page_pre - 1;

	/* last page */
	if(doc_count % rows != 0)
		page_count++;

	/* page range */
	if(page_actual - page_pre < 0)
	{
		page_post += - (page_actual - page_pre);
		page_pre  -= - (page_actual - page_pre);
	}
	else if(page_actual + page_post > page_count)
	{
		page_post = pages - page_pre;
		page_pre  = page_actual + pages - page_count;
	}

	/* sanity */
	if(page_pre < 0)
		page_pre = 0;
	if(page_post < 0)
		page_post = 0;

	/* next pages list */
	int idx = (page_actual - page_pre);
	for(idx = (idx <= 0 ? 0 : idx);
			idx < page_count && idx <= (page_actual + page_post);
				idx++) {
		lst2.add(Integer.toString(idx + 1), idx * rows);
	}
	lst.add("pages", lst2);

	/* navi */
	if(page_actual > 0)
		lst.add("prev", (page_actual - 1) * rows);
	if(page_actual - page_pre > 0)
		lst.add("first", 0);
	if(page_actual < (page_count - 1))
		lst.add("next", (page_actual + 1) * rows);
	if(page_actual + page_post < (page_count - 1))
		lst.add("last", (page_count - 1) * rows);
	lst.add("actual", page_actual + 1);
	lst.add("count", page_count);

	/* finish */
	rb.rsp.add("pager", lst);
  }

  public void modifyRequest(ResponseBuilder rb, SearchComponent who, ShardRequest sreq) 
  {
  }

  @Override
  public void handleResponses(ResponseBuilder rb, ShardRequest sreq) 
  {
  }

  @Override
  public void finishStage(ResponseBuilder rb) 
  {
  }

  ////////////////////////////////////////////
  ///  SolrInfoMBean
  ////////////////////////////////////////////
  
  @Override
  public String getDescription() {
    return "Solr Pager";
  }
  
  @Override
  public String getVersion() {
    return "1.0";
  }

  @Override
  public String getSource() {
    return "URL: https://devel.dob.sk/solr_pager";
  }
  
  @Override
  public URL[] getDocs() {
    return null;
  }
}
